src/search/pdbs/cegar.cc

```diff
@@ -0,0 +1,589 @@
+                break;
+            }
+        }
+
+        // If all equivalent operators are inapplicable, stop plan execution.
+        if (step_failed) {
+            break;
+        }
+    }
+    if (verbosity >= utils::Verbosity::VERBOSE) {
+        if (flaws.empty()) {
+            utils::g_log << "success." << endl;
+        } else {
+            utils::g_log << "failure." << endl;
+        }
+    }
+    return flaws;
+}
+
+FlawList CEGAR::get_flaws_for_projection(
+    int collection_index, const State &concrete_init) {
+    const Projection &projection = *projection_collection[collection_index];
+    if (projection.is_unsolvable()) {
+        utils::g_log << "task is unsolvable." << endl;
+        utils::exit_with(utils::ExitCode::SEARCH_UNSOLVABLE);
+    }
+
+    State current(concrete_init);
+    current.unpack();
+    FlawList flaws = apply_plan(collection_index, current);
+    if (flaws.empty()) {
+        if (task_properties::is_goal_state(task_proxy, current)) {
+            if (verbosity >= utils::Verbosity::VERBOSE) {
+                utils::g_log << "plan led to a concrete goal state: ";
+            }
+            if (blacklisted_variables.empty()) {
+                if (verbosity >= utils::Verbosity::VERBOSE) {
+                    utils::g_log << "there are no blacklisted variables, "
+                        "task solved." << endl;
+                }
+                concrete_solution_index = collection_index;
+            } else {
+                if (verbosity >= utils::Verbosity::VERBOSE) {
+                    utils::g_log << "there are blacklisted variables, "
+                        "marking projection as solved." << endl;
+                }
+                projection_collection[collection_index]->mark_as_solved();
+            }
+        } else {
+            if (verbosity >= utils::Verbosity::VERBOSE) {
+                utils::g_log << "plan did not lead to a goal state: ";
+            }
+            for (const FactPair &goal : goals) {
+                int goal_var_id = goal.var;
+                if (current[goal_var_id].get_pair() != goal && !blacklisted_variables.count(goal_var_id)) {
+                    flaws.emplace_back(collection_index, goal_var_id);
+                }
+            }
+            if (flaws.empty()) {
+                if (verbosity >= utils::Verbosity::VERBOSE) {
+                    utils::g_log << "there are no non-blacklisted goal variables "
+                        "left, marking projection as solved." << endl;
+                }
+                projection_collection[collection_index]->mark_as_solved();
+            } else {
+                if (verbosity >= utils::Verbosity::VERBOSE) {
+                    utils::g_log << "raising goal violation flaw(s)." << endl;
+                }
+            }
+        }
+    }
+    return flaws;
+}
+
+FlawList CEGAR::get_flaws() {
+    FlawList flaws;
+    State concrete_init = task_proxy.get_initial_state();
```

Comment:
Good idea, even though I don't think it matters performance-wise. I just wonder how the copy actually works: if I copy an unpacked state, will the copy be unpacked, too? I couldn't find a copy constructor in `task_proxy.h`.

Reply:
> Good idea, even though I don't think it matters performance-wise.

You're probably right that it doesn't matter, it's just a few pointers and ints that are copied. 

>  I just wonder how the copy actually works: if I copy an unpacked state, will the copy be unpacked, too? I couldn't find a copy constructor in `task_proxy.h`.

I think the copy is unpacked as the implicit copy constructor of the State class calls all copy constructors of its member variables, and the copy consrtuctor of a shared_ptr should create a copy of the shared_ptr. I don't know for sure though.

Reply:
This is my understanding, too.

Reply:
Yes, if there are no user-defined copy constructors, copies are performed elementwise, which in case of a shared pointer means just copying the pointer (and incrementing its reference count), so copying a state is a fairly cheap operation.